File: frontend/src/types/order.ts
// Mirrors the backend API contract for the Bean Tradition ordering system.

export type OrderStatus =
  | 'RECEIVED'
  | 'PREPARING'
  | 'READY'
  | 'DELIVERED'
  | 'CANCELLED';

export type MessageStatus =
  | 'NOT_SENT'
  | 'SEND_PENDING'
  | 'SENT'
  | 'FAILED'
  | 'DELIVERED'
  | 'READ';

export type PaymentMethod = 'UPI' | 'Cash';

export interface OrderItem {
  product: string;
  quantity: number;
}

export interface OrderPublic {
  order_id: string;
  order_number: string;
  created_at: string;
  updated_at: string;
  customer_name: string;
  mobile_masked: string;
  items: OrderItem[];
  items_display: string;
  quantity_total: number;
  notes: string;
  amount: number | null;
  payment_method: string;
  status: OrderStatus;
  confirmation_status: MessageStatus;
  ready_status: MessageStatus;
  ready_at: string | null;
  delivered_at: string | null;
  created_by: string;
  last_messaging_error: string | null;
}

export interface CreateOrderRequest {
  customer_name: string;
  mobile: string;
  items: OrderItem[];
  notes: string;
  amount: number | null;
  payment_method: string;
  created_by: string;
  idempotency_key: string;
}

export interface WhatsAppResult {
  confirmation_sent: boolean;
  meta_message_id: string | null;
  error: string | null;
}

export interface CreatedOrder {
  order_id: string;
  order_number: string;
  status: OrderStatus;
  tracking_url: string;
}

export interface CreateOrderResponse {
  success: boolean;
  order_created: boolean;
  order: CreatedOrder;
  whatsapp: WhatsAppResult;
}

export interface OrdersListResponse {
  orders: OrderPublic[];
  server_time: string;
}

export interface MessageActionResponse {
  success: boolean;
  message_status: MessageStatus;
  meta_message_id: string | null;
  error: string | null;
}

export interface LoginResponse {
  success: boolean;
}

export interface AuthMeResponse {
  authenticated: boolean;
  subject?: string;
}

export interface TrackMyOrder {
  order_number: string;
  status: OrderStatus;
  items_display: string;
}

export interface TrackResponse {
  my_order: TrackMyOrder;
  ready_orders: string[];
  preparing_orders: string[];
  server_time: string;
}

export interface OrderSearchParams {
  order_number?: string;
  mobile_last4?: string;
}

File: backend/app/schemas/order.py
"""Request/response schemas for order creation and management."""
from __future__ import annotations

from pydantic import BaseModel, Field, field_validator

from app.models.order import MessageStatus, OrderInternal, OrderStatus


class OrderItem(BaseModel):
    product: str = Field(min_length=1, max_length=80)
    quantity: int = Field(ge=1, le=99)


class OrderCreate(BaseModel):
    customer_name: str = Field(min_length=1, max_length=120)
    mobile: str = Field(min_length=4, max_length=20)
    items: list[OrderItem] = Field(min_length=1)
    notes: str = Field(default="", max_length=500)
    amount: float | None = Field(default=None, ge=0)
    payment_method: str = Field(default="", max_length=20)
    created_by: str = Field(default="", max_length=60)
    idempotency_key: str = Field(min_length=8, max_length=100)

    @field_validator("customer_name", "notes", "payment_method", "created_by")
    @classmethod
    def _strip(cls, v: str) -> str:
        return v.strip()


class OrderSummary(BaseModel):
    order_id: str
    order_number: str
    status: OrderStatus
    tracking_url: str


class WhatsAppResultBlock(BaseModel):
    confirmation_sent: bool
    meta_message_id: str | None = None
    error: str | None = None


class OrderResponse(BaseModel):
    success: bool
    order_created: bool
    order: OrderSummary
    whatsapp: WhatsAppResultBlock


class OrderStatusUpdate(BaseModel):
    status: OrderStatus
    force: bool = False  # allow reverse/correction transitions


class OrderPublic(BaseModel):
    """Full staff-facing order (safe for authenticated /orders board)."""

    order_id: str
    order_number: str
    created_at: str
    updated_at: str
    customer_name: str
    mobile_masked: str
    items: list[OrderItem]
    items_display: str
    quantity_total: int
    notes: str
    amount: float | None
    payment_method: str
    status: OrderStatus
    confirmation_status: MessageStatus
    ready_status: MessageStatus
    ready_at: str
    delivered_at: str
    created_by: str
    last_messaging_error: str

    @classmethod
    def from_internal(cls, o: OrderInternal) -> "OrderPublic":
        masked = o.mobile[:3] + "XXXXX" + o.mobile[-2:] if len(o.mobile) > 5 else "XXXX"
        return cls(
            order_id=o.order_id,
            order_number=o.order_number,
            created_at=o.created_at,
            updated_at=o.updated_at,
            customer_name=o.customer_name,
            mobile_masked=masked,
            items=[OrderItem(product=i.product, quantity=i.quantity) for i in o.items],
            items_display=o.items_display,
            quantity_total=o.quantity_total,
            notes=o.notes,
            amount=o.amount,
            payment_method=o.payment_method,
            status=o.status,
            confirmation_status=o.confirmation_status,
            ready_status=o.ready_status,
            ready_at=o.ready_at,
            delivered_at=o.delivered_at,
            created_by=o.created_by,
            last_messaging_error=o.last_messaging_error,
        )


class ActiveOrdersResponse(BaseModel):
    orders: list[OrderPublic]
    server_time: str


class MessageActionResponse(BaseModel):
    success: bool
    message_status: MessageStatus
    meta_message_id: str | None = None
    error: str | None = None
